portfolio/src/main/java/com/google/sps/servlets/ServeCommentsServlet.java

```diff
@@ -45,19 +55,50 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String email = request.getParameter("email");
     String content = request.getParameter("content");
-    long timestamp = System.currentTimeMillis();
+    GoogleIdToken idToken = validateUserIdToken(request.getParameter("userTokenId"));
+    
+    if (idToken != null) {
+      System.out.println("Token ID is valid.");
+      addCommentToDb(idToken, content);
+    } else {
+      System.out.println("Invalid ID token.");
+    }
+
+    response.sendRedirect("/leave-comment/leave-comment.html");
+  }
 
+  private void addCommentToDb(GoogleIdToken idToken, String content) {
+    String email = idToken.getPayload().getEmail();
+    long timestamp = System.currentTimeMillis();
     Entity commentEntity = new Entity("Comment");
     commentEntity.setProperty("email", email);
     commentEntity.setProperty("content", content);
     commentEntity.setProperty("timestamp", timestamp);
-
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     datastore.put(commentEntity);
+  }
+
+  private GoogleIdToken validateUserIdToken(String idTokenString) {
+    GoogleIdTokenVerifier verifier = new GoogleIdTokenVerifier.Builder(
+        new NetHttpTransport(),
+        JacksonFactory.getDefaultInstance())
+    .setAudience(Collections.singletonList(CLIENT_ID))
+    .build();
+
+    try {
+      return verifier.verify(idTokenString);
+    } catch (GeneralSecurityException e) {
+      System.out.println(e);
+      return null;
+    } catch (IOException e) {
+      System.out.println(e);
+      return null;
+    } catch (java.lang.NullPointerException e) {
```

Comment:
Why are you catching an NPE here? If you know which method can return a `null` during its "normal" operation, it's better to have a dedicated `if` with a well-defined behavior that will handle that particular case.